### ISSUE TITLE
Add exercise to stress how functions are different from math functions

### ISSUE THREAD

**Issue by iglpdc** [swcarpentry/python-novice-inflammation#489]
Mark Guzdial had [an interesting post a few days ago](https://computinged.wordpress.com/2018/02/26/confusion-over-the-forms-of-programming-problems-mathematics-physics-versus-cs/) on how learners with a math background may assume mistakenly that a function in programming is the same as a function in math. The exercise just asks the learner what's the output of:

```python
def computeSalary():
  rate = 10.00
  hours = 40
  salary = hours * rate
  hours = 30
  print salary
  increase = 1.00
  salary = hours * (rate + increase)
  print salary
```
I tried it and got it wrong and realized that, in fact, I've always thought in functions as mathematical functions.

As we have many learners from science and math camps, I think it would be a good addition to the episode on functions.

**Comment by annefou** [comment]
a will be None because the function one does not return a value. Same for print. 
If there is no return statement in the function code, the function ends, when the control flow reaches the end of the function body and the value "None" will be returned.

**Comment by maxim-belkin** [comment]
:) I [know](https://stackoverflow.com/a/43029507/7410631)

**Comment by iglpdc** [comment]
I think the point of Mark's exercise is not so much distinguishing `return` vs `print`, but the fact that people with a mathematical background tend to assume that functions in programming are equivalent to mathematical functions. A mathematical function will map an input to an output or some input argument and return a value (`None` counts as a value here).

In the exercise, people may follow this trap and say that there is *only one* number being printed. I think not using `return` and instead using `print` twice stresses the issue. A function returns thru one `return` line every time is executed, so it looks like a mathematical function, but in cases like the above we see that is actually a set of statements. 

I think @maxim-belkin 's exercise is too "easy" :)

```
def one():
    print(1)
    print(5)

a = one() # what is the value of `a`?
```
Also the value of `a` is not the point, in the sense that `a` can have only one value. The point is that two numbers are printed (I guess most people would say that here and that's why I'm saying it's easy). 

But in Mark's example, you start doing the math and think that the important thing is that the value of `salary` changes, you only keep track of the updated value and because it really looks like a mathematical function, you may miss the first `print(salary)` if you have the (wrong) preconception that functions are mathematical functions.

@annefou's response might reinforce the wrong preconception: 

> a will be None because the function one does not return a value. Same for print.
If there is no return statement in the function code, the function ends, when the control flow reaches the end of the function body and the value "None" will be returned.

If I have the wrong idea about functions, I'm still confortable with this response because it's focusing in the point that functions take something and return something else.

To summarize, I'd say that the point of the exercise is to stress that despite functions can return things and have only one exit point (so they return one thing), they can execute arbitrary code and should not be thought of as mathematical functions.

**Comment by IgnasiLucas** [comment]
Honestly, I would not be concerned about the possible misconceptions that might derive from a mathematical understanding of functions in programming. In his post, Mark Gudzial acknowledges that "We’d need to do real research to see if there is anything to this, not just me noting anecdotal comments from a half dozen students." Plus, the confusion in his classroom arose from a very specific function that seems to be written to confuse students. In the lesson, functions are properly defined as "a shorthand way of re-executing longer pieces of code". And the motivation of the lesson makes it clear that functions are re-usable pieces of code.

In any case, I think most potential confusions stem from the `return` statement, which may be new for participants, and makes a python function resemble a mathematical function (e.g. `fahr_to_celsius`). The "Return versus print" exercise already helps decouple what a function does from what it returns. To further prevent potential confusions of the sort described by Guzdial, I suggest only one small change: to stress that some functions do not use the `return` statement. Examples of this are already seen in the lesson, like the `analize()` and `detect_problems()` functions. So, it's a matter of adding a sentence like "Notice that not all functions include a return statement".

**Comment by maxim-belkin** [comment]
> I suggest only one small change: to stress that some functions do not use the `return` statement. Examples of this are already seen in the lesson, like the `analize()` and `detect_problems()` functions. So, it's a matter of adding a sentence like "Notice that not all functions include a return statement".

Would you (@IgnasiLucas) be willing to submit a PR?

**Comment by IgnasiLucas** [comment]
Yes. Excited to send my first PR. Give me a couple of days to find the time. Thanks.
